File: reference/src/adapters/interfaces.ts
/**
 * Adapter interface contracts translated from spec/adapters.md.
 * Each interface defines the methods an adapter MUST provide.
 */

// ── Shared Types ──────────────────────────────────────────────────────

/** An async event stream for watch operations. */
export interface EventStream<T> {
  [Symbol.asyncIterator](): AsyncIterator<T>;
}

// ── IssueTracker ──────────────────────────────────────────────────────

export interface IssueFilter {
  status?: string;
  labels?: string[];
  assignee?: string;
  project?: string;
}

export interface Issue {
  id: string;
  title: string;
  description?: string;
  status: string;
  labels?: string[];
  assignee?: string;
  url: string;
}

export interface CreateIssueInput {
  title: string;
  description?: string;
  labels?: string[];
  assignee?: string;
  project?: string;
}

export interface UpdateIssueInput {
  title?: string;
  description?: string;
  labels?: string[];
  assignee?: string;
}

export interface IssueEvent {
  type: 'created' | 'updated' | 'transitioned';
  issue: Issue;
  timestamp: string;
}

export interface IssueComment {
  body: string;
}

export interface IssueTracker {
  listIssues(filter: IssueFilter): Promise<Issue[]>;
  getIssue(id: string): Promise<Issue>;
  createIssue(input: CreateIssueInput): Promise<Issue>;
  updateIssue(id: string, input: UpdateIssueInput): Promise<Issue>;
  transitionIssue(id: string, transition: string): Promise<Issue>;
  addComment(id: string, body: string): Promise<void>;
  getComments(id: string): Promise<IssueComment[]>;
  watchIssues(filter: IssueFilter): EventStream<IssueEvent>;
}

// ── SourceControl ─────────────────────────────────────────────────────

export interface CreateBranchInput {
  name: string;
  from?: string;
}

export interface Branch {
  name: string;
  sha: string;
}

export interface CreatePRInput {
  title: string;
  description?: string;
  sourceBranch: string;
  targetBranch: string;
}

export type MergeStrategy = 'merge' | 'squash' | 'rebase';

export interface PullRequest {
  id: string;
  title: string;
  description?: string;
  sourceBranch: string;
  targetBranch: string;
  status: 'open' | 'merged' | 'closed';
  author: string;
  url: string;
}

export interface MergeResult {
  sha: string;
  merged: boolean;
}

export interface FileContent {
  path: string;
  content: string;
  encoding: string;
}

export interface ChangedFile {
  path: string;
  status: 'added' | 'modified' | 'deleted' | 'renamed';
  additions: number;
  deletions: number;
}

export interface CommitStatus {
  state: 'pending' | 'success' | 'failure' | 'error';
  context: string;
  description?: string;
  targetUrl?: string;
}

export interface PRFilter {
  status?: string;
  author?: string;
  targetBranch?: string;
}

export interface PREvent {
  type: 'opened' | 'updated' | 'merged' | 'closed';
  pullRequest: PullRequest;
  timestamp: string;
}

export interface SourceControl {
  createBranch(input: CreateBranchInput): Promise<Branch>;
  createPR(input: CreatePRInput): Promise<PullRequest>;
  mergePR(id: string, strategy: MergeStrategy): Promise<MergeResult>;
  getFileContents(path: string, ref: string): Promise<FileContent>;
  listChangedFiles(prId: string): Promise<ChangedFile[]>;
  setCommitStatus(sha: string, status: CommitStatus): Promise<void>;
  watchPREvents(filter: PRFilter): EventStream<PREvent>;
}

// ── CIPipeline ────────────────────────────────────────────────────────

export interface TriggerBuildInput {
  branch: string;
  commitSha?: string;
  parameters?: Record<string, string>;
}

export interface Build {
  id: string;
  status: string;
  url?: string;
}

export interface BuildStatus {
  id: string;
  status: 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled';
  startedAt?: string;
  completedAt?: string;
}

export interface TestResults {
  passed: number;
  failed: number;
  skipped: number;
  duration?: number;
}

export interface CoverageReport {
  lineCoverage: number;
  branchCoverage?: number;
  functionCoverage?: number;
}

export interface BuildFilter {
  branch?: string;
  status?: string;
}

export interface BuildEvent {
  type: 'started' | 'completed' | 'failed';
  build: Build;
  timestamp: string;
}

export interface CIPipeline {
  triggerBuild(input: TriggerBuildInput): Promise<Build>;
  getBuildStatus(id: string): Promise<BuildStatus>;
  getTestResults(buildId: string): Promise<TestResults>;
  getCoverageReport(buildId: string): Promise<CoverageReport>;
  watchBuildEvents(filter: BuildFilter): EventStream<BuildEvent>;
}

// ── CodeAnalysis ──────────────────────────────────────────────────────

export interface ScanInput {
  repository: string;
  branch?: string;
  commitSha?: string;
  rulesets?: string[];
}

export interface ScanResult {
  id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
}

export type Severity = 'low' | 'medium' | 'high' | 'critical';

export interface Finding {
  id: string;
  severity: Severity;
  message: string;
  file: string;
  line?: number;
  rule: string;
}

export interface SeveritySummary {
  critical: number;
  high: number;
  medium: number;
  low: number;
}

export interface CodeAnalysis {
  runScan(input: ScanInput): Promise<ScanResult>;
  getFindings(scanId: string): Promise<Finding[]>;
  getSeveritySummary(scanId: string): Promise<SeveritySummary>;
}

// ── Messenger ─────────────────────────────────────────────────────────

export interface NotificationInput {
  channel: string;
  message: string;
  severity?: 'info' | 'warning' | 'error';
}

export interface ThreadInput {
  channel: string;
  title: string;
  message: string;
}

export interface Thread {
  id: string;
  url: string;
}

export interface Messenger {
  sendNotification(input: NotificationInput): Promise<void>;
  createThread(input: ThreadInput): Promise<Thread>;
  postUpdate(threadId: string, message: string): Promise<void>;
}

// ── DeploymentTarget ──────────────────────────────────────────────────

export interface DeployInput {
  artifact: string;
  environment: string;
  version: string;
  parameters?: Record<string, string>;
}

export interface Deployment {
  id: string;
  status: string;
  environment: string;
  url?: string;
}

export interface DeploymentStatus {
  id: string;
  status: 'pending' | 'in-progress' | 'succeeded' | 'failed' | 'rolled-back';
  environment: string;
  timestamp: string;
}

export interface DeployFilter {
  environment?: string;
  status?: string;
}

export interface DeployEvent {
  type: 'started' | 'succeeded' | 'failed' | 'rolled-back';
  deployment: Deployment;
  timestamp: string;
}

export interface DeploymentTarget {
  deploy(input: DeployInput): Promise<Deployment>;
  getDeploymentStatus(id: string): Promise<DeploymentStatus>;
  rollback(id: string): Promise<Deployment>;
  watchDeploymentEvents(filter: DeployFilter): EventStream<DeployEvent>;
}

// ── Adapter Map ───────────────────────────────────────────────────────

export interface AdapterInterfaces {
  IssueTracker: IssueTracker;
  SourceControl: SourceControl;
  CIPipeline: CIPipeline;
  CodeAnalysis: CodeAnalysis;
  Messenger: Messenger;
  DeploymentTarget: DeploymentTarget;
}

File: sdk-go/adapters/interfaces.go
// Package adapters provides adapter interfaces, registry, and implementations.
package adapters

import "context"

// Issue represents an issue from an issue tracker.
type Issue struct {
	ID          string            `json:"id"`
	Title       string            `json:"title"`
	Description string            `json:"description,omitempty"`
	Status      string            `json:"status"`
	Assignee    string            `json:"assignee,omitempty"`
	Labels      []string          `json:"labels,omitempty"`
	URL         string            `json:"url,omitempty"`
	Metadata    map[string]string `json:"metadata,omitempty"`
}

// IssueTracker manages issues.
type IssueTracker interface {
	GetIssue(ctx context.Context, id string) (*Issue, error)
	CreateIssue(ctx context.Context, issue *Issue) (*Issue, error)
	UpdateIssue(ctx context.Context, id string, updates map[string]interface{}) (*Issue, error)
	AddComment(ctx context.Context, issueID, comment string) error
	ListIssues(ctx context.Context, filter map[string]string) ([]*Issue, error)
}

// PullRequest represents a pull/merge request.
type PullRequest struct {
	ID          string   `json:"id"`
	Title       string   `json:"title"`
	Description string   `json:"description,omitempty"`
	SourceBranch string  `json:"sourceBranch"`
	TargetBranch string  `json:"targetBranch"`
	Status      string   `json:"status"`
	URL         string   `json:"url,omitempty"`
}

// SourceControl manages source code repositories.
type SourceControl interface {
	CreateBranch(ctx context.Context, repo, branchName, fromRef string) error
	CreatePullRequest(ctx context.Context, pr *PullRequest) (*PullRequest, error)
	GetPullRequest(ctx context.Context, repo, id string) (*PullRequest, error)
	MergePullRequest(ctx context.Context, repo, id string) error
	GetFileContent(ctx context.Context, repo, path, ref string) ([]byte, error)
}

// PipelineRun represents a CI pipeline execution.
type PipelineRun struct {
	ID     string `json:"id"`
	Status string `json:"status"`
	URL    string `json:"url,omitempty"`
}

// CIPipeline manages CI/CD pipelines.
type CIPipeline interface {
	TriggerPipeline(ctx context.Context, repo, ref string, params map[string]string) (*PipelineRun, error)
	GetPipelineStatus(ctx context.Context, repo, runID string) (*PipelineRun, error)
	CancelPipeline(ctx context.Context, repo, runID string) error
}

// AnalysisFinding represents a code analysis finding.
type AnalysisFinding struct {
	RuleID   string `json:"ruleId"`
	Severity string `json:"severity"`
	Message  string `json:"message"`
	File     string `json:"file"`
	Line     int    `json:"line"`
}

// AnalysisResult is the result of a code analysis run.
type AnalysisResult struct {
	Tool     string            `json:"tool"`
	Passed   bool              `json:"passed"`
	Findings []AnalysisFinding `json:"findings"`
}

// CodeAnalysis runs static analysis on code.
type CodeAnalysis interface {
	Analyze(ctx context.Context, repo, ref string, rulesets []string) (*AnalysisResult, error)
}

// Message represents a notification message.
type Message struct {
	Target  string `json:"target"`
	Title   string `json:"title"`
	Body    string `json:"body"`
	Channel string `json:"channel,omitempty"`
}

// Messenger sends notifications.
type Messenger interface {
	Send(ctx context.Context, msg *Message) error
}

// Deployment represents a deployment.
type Deployment struct {
	ID          string `json:"id"`
	Environment string `json:"environment"`
	Status      string `json:"status"`
	URL         string `json:"url,omitempty"`
}

// DeploymentTarget manages deployments.
type DeploymentTarget interface {
	Deploy(ctx context.Context, environment, artifact string, config map[string]string) (*Deployment, error)
	GetDeploymentStatus(ctx context.Context, deploymentID string) (*Deployment, error)
	Rollback(ctx context.Context, deploymentID string) error
}

// Event represents an event in the event bus.
type Event struct {
	Type    string                 `json:"type"`
	Source  string                 `json:"source"`
	Data    map[string]interface{} `json:"data"`
}

// EventHandler processes events.
type EventHandler func(ctx context.Context, event *Event) error

// EventBus provides publish/subscribe messaging.
type EventBus interface {
	Publish(ctx context.Context, event *Event) error
	Subscribe(eventType string, handler EventHandler) (func(), error)
}
